src/platform/storage/zephyr/nvs/storage.c

```diff
@@ -163,9 +197,149 @@ mender_storage_init(void) {
+            mender_log_error("Failed to rotate the deployment logs FCB");
+            /* if (0 != (result = fcb_append_to_scratch(&depl_logs_buffer))) { */
+            /*     mender_log_error("Failed to start using the scratch sector for the deployment logs FCB"); */
+            ret = MENDER_FAIL;
+            goto END;
+        }
+        const char rotation_msg[] = "<wrn> ------- DEPLOYMENT LOGS ROTATED -------";
+        result                    = fcb_append(&depl_logs_buffer, sizeof(rotation_msg), &entry);
+        if (0 != (result = flash_area_write(fap, entry.fe_sector->fs_off + entry.fe_data_off, rotation_msg, sizeof(rotation_msg)))) {
+            mender_log_error("Failed to write message to the FCB entry [%d]", -result);
+            ret = MENDER_FAIL;
+            goto END;
+        }
+        result = fcb_append_finish(&depl_logs_buffer, &entry);
+        if (0 != result) {
+            mender_log_error("Failed to finish append of a rotation entry to the deployment logs FCB");
+            ret = MENDER_FAIL;
+            goto END;
+        }
+
+        result = fcb_append(&depl_logs_buffer, msg_size, &entry);
+    }
+    if (0 != result) {
+        mender_log_error("Failed to append a new entry to the deployment logs FCB");
+        ret = MENDER_FAIL;
+        goto END;
+    }
+    /* else success, proceed */
+    if (0 != (result = flash_area_write(fap, entry.fe_sector->fs_off + entry.fe_data_off, msg, msg_size))) {
+        mender_log_error("Failed to write message to the FCB entry [%d]", -result);
+        ret = MENDER_FAIL;
+        goto END;
+    }
+    result = fcb_append_finish(&depl_logs_buffer, &entry);
+    if (0 != result) {
+        mender_log_error("Failed to finish append of a new entry to the deployment logs FCB");
+        ret = MENDER_FAIL;
+        goto END;
+    }
+
+END:
+    flash_area_close(fap);
+
+    return ret;
+}
+
+mender_err_t
+mender_storage_deployment_log_walk(MenderDeploymentLogVisitor visitor_fn, void *ctx) {
+    mender_err_t ret = MENDER_OK;
+    int          result;
+    char         msg[DEPL_LOGS_MAX_MSG_LEN + 1];
+
+    const struct flash_area *fap;
+    if (0 != (result = flash_area_open(FIXED_PARTITION_ID(MENDER_STORAGE_LABEL), &fap))) {
+        mender_log_error("Unable to open the Mender storage flash area");
+        return MENDER_FAIL;
+    }
+
+    struct fcb_entry entry = { 0 };
+    result                 = fcb_getnext(&depl_logs_buffer, &entry);
+    while (0 == result) {
+        if (0 != (result = flash_area_read(fap, entry.fe_sector->fs_off + entry.fe_data_off, (void *)msg, MIN(DEPL_LOGS_MAX_MSG_LEN, entry.fe_data_len)))) {
+            mender_log_error("Failed to read FCB entry from flash [%d]", -result);
+            ret = MENDER_FAIL;
+            goto END;
+        }
+        msg[MIN(DEPL_LOGS_MAX_MSG_LEN, entry.fe_data_len)] = '\0';
+        visitor_fn(msg, ctx);
+        result = fcb_getnext(&depl_logs_buffer, &entry);
```

Comment:
If `fcb_getnext` gives any error it will break the loop. The intention is to break on "no more entries" error, but we would be masking any other error too.

We should check outside of the loop that result is "no more entries" error.